pysmt/rewritings.py

```diff
@@ -673,6 +679,113 @@ def walk_minus(self, formula, args, **kwargs):
 # EOC TimesDistributivity
 
 
+
+class Ackermannizer(IdentityDagWalker):
+    def __init__(self, environment=None):
+        IdentityDagWalker.__init__(self, environment)
+        #funs_to_args keeps for every function symbol f,
+        #a set of lists of arguments.
+        #if f(g(x),y) and f(x,g(y)) occur in a formula, then we
+        #will have "f": set([g(x), y], [x, g(y)])
+        self._funs_to_args = {}
+
+        #maps the actual applications to the constants that will be
+        #generated, or to the original term if it is not replaced.
+        self._terms_dict = {}
+
+    def do_ackermannization(self, formula):
+        self._fill_maps(formula)
+        #new formula without function symbols
+        substitued_formula = self._make_substitutions(formula)
+        #function consistency
+        implications = self._get_equality_implications()
+        function_consistency = And(implications)
+        if (len(implications) == 0):
+            result = substitued_formula
+        else:
+            result = And(function_consistency, substitued_formula)
+        return result
+
+    def _get_equality_implications(self):
+        result = set([])
+        for f in self._funs_to_args:
+            implications = self._generate_implications(f)
+            result.update(implications)
+        return result
+
+
+    def _generate_implications(self, f):
+        result = set([])
+        possible_args = self._funs_to_args[f]
+        for option1, option2 in pairwise(possible_args):
+            implication = self._generate_implication(option1, option2, f)
+            result.add(implication)
+        return result
+
+    def _generate_implication(self, option1, option2, f):
+        left_conjuncts = set([])
+        for i in range(0, len(option1)):
+            if option1[i].is_function_application():
+                term1 = self._terms_dict[option1[i]]
+            else:
+                term1 = option1[i]
+            if option2[i].is_function_application():
+                term2 = self._terms_dict[option2[i]]
+            else:
+                term2 = option2[i]
+            conjunct = EqualsOrIff(term2, term2)
+            left_conjuncts.add(conjunct)
+        left = And(left_conjuncts)
+        app1 = Function(f, option1)
+        app2 = Function(f, option2)
+        app1_const = self._terms_dict[app1]
+        app2_const = self._terms_dict[app2]
+        right = EqualsOrIff(app1_const, app2_const)
+        implication = Implies(left, right)
+        return implication
+
+
+
+
+    def _make_substitutions(self, formula):
+        return formula.substitute(self._terms_dict)
+
+    def _fill_maps(self, formula):
+        self.walk(formula)
+
+
+    def walk_function(self, formula, args, **kwargs):
```

Comment:
These are two different issues:
1. You want to make the `_terms_dict` map available to the clients of this class. That is probably a good idea. I would just recommend that you either rename to avoid the leading underscore or (maybe better) you wrap it into a getter method. At the same time, I think you might be more interested in returning a map Symbol -> Term, rather than the current Term->Symbol, but either way is fine with me.

2. What I am trying to say is that you are using the `.substitute` method. This means that you need to navigate the entire formula once to build the map, and then again to do the substitution. This is unnecessary, since you can perform the substitution already within the `walk_function` call. Here is a simple way to fix the method:

```python
def walk_function(self, formula, args, **kwargs):
    rewritten = IdentityDagWalker.super(self,formula, args, **kwargs)
    if rewritten not in self._terms_dict:
        function_name = formula.function_name()
        arguments = formula.args()
        self._add_args_to_fun(function_name, arguments)
        self._add_application(rewritten)
    return self._terms_dict[rewritten]
```

A slightly more pythonic way:
```python
def walk_function(self, formula, args, **kwargs):
    rewritten = IdentityDagWalker.super(self,formula, args, **kwargs)
    try:
        ack_symbol = self._terms_dict[rewritten]
    except KeyError:
        function_name = formula.function_name()
        arguments = formula.args()
        self._add_args_to_fun(function_name, arguments)
        self._add_application(rewritten)
        ack_symbol = self._terms_dict[rewritten]
    return ack_symbol
```

Reply:
Right, both goals could be achieved.
I pushed a variant of your second proposal: instead of adding "rewritten" to the maps, I am adding the original formula. Otherwise, I get symbols that were introduced by Ackermannization as keys in the maps, but I think only original terms should appear as keys, right?